### PR TITLE
feat(sim): update sim crate to use alloy types

### PR DESCRIPTION
**Pull request by dancoombs**

### CODE REVIEW

File: crates/sim/src/gas/oracle.rs [hunk 1]
```diff
@@ -31,20 +31,49 @@ pub(crate) enum FeeOracleError {
...
-pub(crate) trait FeeOracle: Send + Sync + Debug {
-    async fn estimate_priority_fee(&self) -> Result<U256>;
+#[auto_impl::auto_impl(Arc)]
+pub trait FeeOracle: Send + Sync {
+    /// Estimate the priority fee
+    async fn estimate_priority_fee(&self) -> Result<u128>;
+}
+
+/// Get a fee oracle for the given chain spec.
+pub fn get_fee_oracle<'a, P>(chain_spec: &ChainSpec, provider: P) -> Arc<dyn FeeOracle + 'a>
```
Comment (dancoombs) [inline crates/sim/src/gas/oracle.rs:43]: Modify this to return `Box`, the caller can wrap in `Arc` if needed

File: crates/sim/src/gas/oracle.rs [hunk 2]
```diff
@@ -31,20 +31,49 @@ pub(crate) enum FeeOracleError {
     Other(#[from] anyhow::Error),
 }
 
+#[async_trait::async_trait]
 /// FeeOracle is a trait that provides a way to estimate the priority fee
-#[async_trait]
-pub(crate) trait FeeOracle: Send + Sync + Debug {
-    async fn estimate_priority_fee(&self) -> Result<U256>;
+#[auto_impl::auto_impl(Arc)]
```
Comment (dancoombs) [inline crates/sim/src/gas/oracle.rs:36]: Add more impls to this

File: crates/sim/src/simulation/mod.rs [hunk 3]
```diff
@@ -83,6 +84,8 @@ impl SimulationResult {
     }
 }
 
+// TODO(danc): clean up these error types/return types
```
Comment (dancoombs) [inline crates/sim/src/simulation/mod.rs:87]: TODO

File: crates/sim/src/estimation/v0_6.rs [hunk 4]
```diff
@@ -328,16 +326,15 @@ where
             }
         }
 
-        let call_gas_limit: U256 = self
+        let call_gas_limit: u128 = self
```
Comment (dancoombs) [inline crates/sim/src/estimation/v0_6.rs:329]: Remove u128

**Review by andysim3d — APPROVED** [review]

File: crates/provider/src/alloy/entry_point/v0_6.rs [hunk 5]
```diff
@@ -517,7 +517,7 @@ impl From<DepositInfoV0_6> for DepositInfo {
         Self {
             deposit: U256::from(deposit_info.deposit),
             staked: deposit_info.staked,
-            stake: UintTryTo::<u128>::uint_try_to(&deposit_info.stake).unwrap(),
+            stake: U256::from(deposit_info.stake),
```
Comment (dancoombs) [inline crates/provider/src/alloy/entry_point/v0_6.rs:520]: Moved everything representing a token value to a `U256`. We can consider moving this down to a `u128` in the future (which I think should be fine), but for simplicity keeping everything as a `U256` for now.

File: crates/sim/src/estimation/estimate_call_gas.rs [hunk 6]
```diff
@@ -129,68 +121,89 @@ where
...
-                .call_spoofed_simulate_op(
+                .simulate_handle_op(
                     callless_op.clone(),
-                    self.entry_point.address(),
+                    *self.entry_point.address(),
                     target_call_data,
-                    block_hash,
-                    self.settings.max_simulate_handle_ops_gas.into(),
-                    &state_override,
+                    BlockId::Hash(block_hash.into()),
```
Comment (dancoombs) [inline crates/sim/src/estimation/estimate_call_gas.rs:128]: nit: should be able to just do `block_hash.into()`

File: crates/sim/src/estimation/estimate_call_gas.rs [hunk 7]
```diff
@@ -129,68 +121,89 @@ where
...
+                CallGasEstimationProxyErrors::EstimateCallGasContinuation(continuation) => {
+                    let ret_min_gas = continuation
+                        .minGas
+                        .try_into()
+                        .context("min gas return overflow")?;
+                    let ret_max_gas = continuation
+                        .maxGas
+                        .try_into()
+                        .context("max gas return overflow")?;
+                    let ret_num_rounds: u32 = continuation
```
Comment (dancoombs) [inline crates/sim/src/estimation/estimate_call_gas.rs:173]: nit: shouldn't need this `u32`

File: crates/sim/src/estimation/estimate_call_gas.rs [hunk 8]
```diff
@@ -200,30 +213,30 @@ where
...
-            .call_spoofed_simulate_op(
+            .simulate_handle_op(
                 callless_op,
-                self.entry_point.address(),
+                *self.entry_point.address(),
                 target_call_data,
-                block_hash,
-                self.settings.max_simulate_handle_ops_gas.into(),
-                &state_override,
+                BlockId::Hash(block_hash.into()),
```
Comment (dancoombs) [inline crates/sim/src/estimation/estimate_call_gas.rs:220]: nit: should be able to do `block_hash.into()`

File: crates/sim/src/estimation/estimate_verification_gas.rs [hunk 9]
```diff
@@ -128,24 +114,20 @@ where
...
                 .entry_point
-                .call_spoofed_simulate_op(
+                .simulate_handle_op(
                     op,
-                    Address::zero(),
+                    Address::ZERO,
                     Bytes::new(),
-                    block_hash,
-                    self.settings.max_simulate_handle_ops_gas.into(),
+                    BlockId::Hash(block_hash.into()),
```
Comment (dancoombs) [inline crates/sim/src/estimation/estimate_verification_gas.rs:129]: nit: `block_hash.into()`

File: crates/sim/src/gas/oracle.rs [hunk 10]
```diff
@@ -114,20 +142,21 @@ where
...
         }
 
+        let Some(reward) = fee_history.reward else {
+            return Ok(0);
+        };
+
         // Chain is congested, return the average of the last N blocks at configured percentile
-        let values = fee_history
-            .reward
+        let values = reward
```
Comment (dancoombs) [inline crates/sim/src/gas/oracle.rs:150]: Re-check this logic

File: crates/sim/src/gas/oracle.rs [hunk 11]
```diff
@@ -141,69 +170,72 @@ pub(crate) struct FeeHistoryOracleConfig {
...
+        let fee = calculate_estimate_from_rewards(&reward);
         Ok(fee.clamp(self.config.minimum_fee, self.config.maximum_fee))
     }
 }
 
 // Calculates the estimate based on the index of inner vector
 // and skips the average if block is empty
-fn calculate_estimate_from_rewards(reward: &[Vec<U256>]) -> U256 {
+fn calculate_estimate_from_rewards(reward: &[Vec<u128>]) -> u128 {
     let mut values = reward
```
Comment (dancoombs) [inline crates/sim/src/gas/oracle.rs:232]: Recheck this math, do we need it twice?

File: crates/sim/src/gas/oracle.rs [hunk 12]
```diff
@@ -262,14 +292,14 @@ impl MaxOracle {
         Self { oracles: vec![] }
     }
 
-    pub(crate) fn add<T: FeeOracle + 'static>(&mut self, oracle: T) {
-        self.oracles.push(Box::new(oracle));
+    pub(crate) fn add(&mut self, oracle: Box<dyn FeeOracle>) {
+        self.oracles.push(oracle);
```
Comment (dancoombs) [inline crates/sim/src/gas/oracle.rs:296]: Was this change needed?

File: crates/sim/src/simulation/unsafe_sim.rs [hunk 13]
```diff
@@ -79,36 +77,28 @@ where
...
 
         // simulate the validation
         let validation_result = self
             .entry_point
-            .call_simulate_validation(
+            .simulate_validation(
                 op.clone(),
                 self.sim_settings.max_verification_gas,
-                Some(block_hash),
+                Some(BlockId::Hash(block_hash.into())),
```
Comment (dancoombs) [inline crates/sim/src/simulation/unsafe_sim.rs:90]: nit: `block_hash.into()`

File: crates/sim/src/simulation/v0_6/context.rs [hunk 14]
```diff
@@ -155,31 +159,27 @@ where
...
         // crafted UO can use extra verification gas that isn't caught during simulation, but when
         // it runs on chain causes the transaction to revert.
-        let verification_gas_used = entry_point_out
-            .return_info
-            .pre_op_gas
-            .saturating_sub(op.pre_verification_gas());
         let verification_buffer = op
             .total_verification_gas_limit()
-            .saturating_sub(verification_gas_used);
+            .saturating_sub(entry_point_out.return_info.pre_op_gas);
```
Comment (dancoombs) [inline crates/sim/src/simulation/v0_6/context.rs:164]: This looks wrong, should have subtracted PVG?

File: crates/sim/src/simulation/v0_6/context.rs [hunk 15]
```diff
@@ -155,31 +159,27 @@ where
...
         if verification_buffer < REQUIRED_VERIFICATION_GAS_LIMIT_BUFFER {
             violations.push(SimulationViolation::VerificationGasLimitBufferTooLow(
                 op.total_verification_gas_limit(),
-                verification_gas_used + REQUIRED_VERIFICATION_GAS_LIMIT_BUFFER,
+                entry_point_out.return_info.pre_op_gas + REQUIRED_VERIFICATION_GAS_LIMIT_BUFFER,
```
Comment (dancoombs) [inline crates/sim/src/simulation/v0_6/context.rs:168]: Looks wrong

File: crates/sim/src/simulation/v0_7/context.rs [hunk 16]
```diff
@@ -272,31 +263,16 @@ impl<T> ValidationContextProvider<T> {
...
-                } else {
-                    Ok(Err(ValidationRevert::Unknown(
-                        Bytes::from_hex(top.exit_data.clone())
-                            .context("failed to parse exit data has hex")?,
-                    )))
-                }
+                let bytes: Bytes = hex::decode(top.exit_data.clone())
+                    .context("should decode hex exit data")?
+                    .into();
+                Ok(Err(rundler_provider::decode_v0_7_validation_revert(&bytes)))
```
Comment (dancoombs) [inline crates/sim/src/simulation/v0_7/context.rs:269]: Double check this
